source/_posts/2018/2018-02-19-rector-part-1-what-and-how.md

```diff
@@ -0,0 +1,145 @@
+
+From this:
+
+```php
+class LectureController extends BaseController
+{
+    public function listAction()
+    {
+        $logger = $this->get('logger');
+        $logger->log('it happened!');
+    }
+}
+```
+
+To this:
+
+```php
+class LectureController extends BaseController
+{
+    /**
+     * @var LoggerInterface
+     */ 
+    private $loggerInterface;
+    
+    public function __construct(LoggerInterface $loggerInterface)
+    {
+        $this->loggerInterface = $loggerInterface;
+    }
+
+    public function listAction()
+    {
+        $this->logger->log('it happened!');
+    }
+}
+```
+
+This can Rector handle too.
+
+Do you use Laravel and want to move from facades to constructor injection? Rector can help you. 
+
+## How it Works?
+
+Rector parses the code to [AST](/blog/2017/11/06/wow-to-change-php-code-with-abstract-syntax-tree/) thanks to PHP superman [nikic](https://nikic.github.io/)'s [php-parser](https://github.com/nikic/PHP-Parser).
+
+Then it finds specific place in the code, e.g. all variables that containt `Symfony\Component\HttpFoundation\Request` type and call 'isMethodSafe()' method.
+ 
+Then it changes it into 'isMethodCacheable()' (see [UPGRADE-4.0.md](https://github.com/symfony/symfony/blob/master/UPGRADE-4.0.md#httpfoundation)).
+
+Such configuration looks like this (as shown in [`README`](https://github.com/rectorphp/rector#change-a-method-name)):
+
+```yaml
+# rector.yml
+rectors:
+    # prepared service that handles method name changes
+    Rector\Rector\Dynamic\MethodNameReplacerRector:
+        # type to look for
+        'Symfony\Component\HttpFoundation\Request':
+            # old method name: new method name
+            'isMethodSafe': 'isMethodCacheable'
+```
+
+### Member of Big AST PHP Family
+
+Rector is not the only one who uses `nikic\php-parser` for context-aware operation on your code.
+
+You probably heard of [PHPStan](/blog/2017/01/28/why-I-switched-scrutinizer-for-phpstan-and-you-should-too/). But unfortunately it's read-only for deterministic cases = when 1 error has exactly 1 possible solution. 
+
+A bit further is another static analysis tool - [`vimeo/psalm`](https://github.com/vimeo/psalm) by [Matthew Brown](https://github.com/muglug), which fixes such code. Great job Matthew!
+
+## Easter Egg: Has Google Own "Rector"?
+
+This *setup and forget* approach is so addictive, that Google must have it too, right?
+
+And it does! I found 4-page case study *[Large-Scale Automated Refactoring Using ClangMR](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/41342.pdf)*, ) that was [presented by *Hyrum Wright* on CppCon2014](https://www.youtube.com/watch?v=ZpvvmvITOrk) in 57 minutes. Hyrum doesn't work at Google anymore (as he wrote me), yet I still love his detailed and practical talk. 
+
```

Comment:
I love this article, :+1: for put it here, and thanks for this YouTube link, I didn't know it. Also, I believe there's an extra `)` after the `.pdf`